flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java

```diff
@@ -21,519 +21,897 @@
+			}
 		}
 
-		this.taskManagers.put(resourceID, registration);
+		return null;
+	}
+
+	/**
+	 * Finds a matching slot for a given resource profile. A matching slot has at least as many
+	 * resources available as the given resource profile. If there is no such slot available, then
+	 * the method returns null.
+	 *
+	 * Note: If you want to change the behaviour of the slot manager wrt slot allocation and
+	 * request fulfillment, then you should override this method.
+	 *
+	 * @param requestResourceProfile specifying the resource requirements for the a slot request
+	 * @return A matching slot which fulfills the given resource profile. Null if there is no such
+	 * slot available.
+	 */
+	protected TaskManagerSlot findMatchingSlot(ResourceProfile requestResourceProfile) {
+		Iterator<Map.Entry<SlotID, TaskManagerSlot>> iterator = freeSlots.entrySet().iterator();
+
+		while (iterator.hasNext()) {
+			TaskManagerSlot taskManagerSlot = iterator.next().getValue();
 
-		for (SlotStatus slotStatus : slotReport.getSlotsStatus()) {
-			final SlotID slotId = slotStatus.getSlotID();
+			// sanity check
+			Preconditions.checkState(taskManagerSlot.isFree());
 
-			final TaskExecutorRegistration taskExecutorRegistration = taskManagers.get(slotId.getResourceID());
-			if (taskExecutorRegistration == null) {
-				LOG.info("Received SlotStatus but ResourceID {} is unknown to the SlotManager",
-					slotId.getResourceID());
-				return;
+			if (taskManagerSlot.getResourceProfile().isMatching(requestResourceProfile)) {
+				iterator.remove();
+				return taskManagerSlot;
 			}
+		}
 
-			final ResourceSlot slot = new ResourceSlot(slotId, slotStatus.getProfiler(), taskExecutorRegistration);
+		return null;
+	}
 
-			registerNewSlot(slot);
-			LOG.info("New slot appeared, SlotID:{}, AllocationID:{}", slotId, slotStatus.getAllocationID());
+	// ---------------------------------------------------------------------------------------------
+	// Internal slot operations
+	// ---------------------------------------------------------------------------------------------
 
-			if (slotStatus.getAllocationID() != null) {
-				// slot in use, record this in bookkeeping
-				allocationMap.addAllocation(slotId, slotStatus.getAllocationID());
-			} else {
+	/**
+	 * Registers a slot for the given task manager at the slot manager. The task manager is
+	 * identified by the given instance id and the slot is identified by the given slot id. The
+	 * given resource profile defines the available resources for the slot. The task manager
+	 * connection can be used to communicate with the task manager.
+	 *
+	 * @param instanceId identifying the task manager on which the slot lives
+	 * @param slotId identifying the slot on the task manager
+	 * @param allocationId which is currently deployed in the slot
+	 * @param resourceProfile of the slot
+	 * @param taskManagerConnection to communicate with the remote task manager
+	 */
+	private void registerSlot(
```

Comment:
I would pass the `TaskManagerRegistration` into this method. Looking it up again is redundant work and leads to the "what if it is not found" case (where the question comes up whether to log or to throw an exception), which can be avoided.